src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamService.java

```diff
@@ -236,25 +236,29 @@ else if (exercise instanceof ModelingExercise) {
     }
 
     /**
-     * Assess the modeling-, text and file upload exercises of student exams of an exam which are not submitted with 0 points.
+     * Assess all exercises, except quiz exercises, of student exams of an exam which are not submitted with 0 points.
      *
      * @param exam the exam
      * @param assessor the assessor should be the instructor making the call.
      * @return returns the set of unsubmitted StudentExams, the participations of which were assessed
      */
     public Set<StudentExam> assessUnsubmittedStudentExams(final Exam exam, final User assessor) {
-        // TODO Simon Entholzer: we should also do this for programming exercises with manual assessment
         Set<StudentExam> unsubmittedStudentExams = studentExamRepository.findAllUnsubmittedWithExercisesByExamId(exam.getId());
-        Map<User, List<Exercise>> exercisesOfUser = unsubmittedStudentExams.stream()
-                .collect(Collectors.toMap(StudentExam::getUser,
-                        studentExam -> studentExam.getExercises().stream()
-                                .filter(exercise -> exercise instanceof ModelingExercise || exercise instanceof TextExercise || exercise instanceof FileUploadExercise)
-                                .collect(Collectors.toList())));
+        Map<User, List<Exercise>> exercisesOfUser = getExercisesOfUserMap(unsubmittedStudentExams);
         for (final var user : exercisesOfUser.keySet()) {
+            // fetch all studentParticipations of a user, with sumbissions and results eagerly loaded
             final var studentParticipations = studentParticipationRepository.findByStudentIdAndIndividualExercisesWithEagerSubmissionsResultIgnoreTestRuns(user.getId(),
                     exercisesOfUser.get(user));
+
             for (final var studentParticipation : studentParticipations) {
-                final var latestSubmission = studentParticipation.findLatestSubmission();
+                var latestSubmission = studentParticipation.findLatestSubmission();
+                if (latestSubmission.isEmpty() && studentParticipation.getExercise() instanceof ProgrammingExercise
+                        && ((ProgrammingExercise) studentParticipation.getExercise()).areManualResultsAllowed()) {
+                    // when it is the participation of a programming exercise and manual assessment is enabled, but there is no submission,
+                    // a new submission for the programming participation needs to be created
+                    submissionService.addEmptyProgrammingSubmissionToParticipation(studentParticipation);
+                    latestSubmission = studentParticipation.findLatestSubmission();
+                }
```

Comment:
This part appears twice

Reply:
I agree, please try to extract a method.

Reply:
I extracted it to a smaller helper method :+1: